### PR TITLE
Add on_exception callback hook

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?
Create an `on_exception` callback hook that users can override to implement custom behavior when an exception occurs.

Addresses #8992

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -500,20 +500,23 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
...
+                self.on_exception(exception)
+        except BaseException as exception:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
                 # try syncing remaing processes, kill otherwise
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             self._on_exception()
             # reset bookkeeping
             self.state.stage = None
+            self.on_exception(exception)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:519]: it's not something we can solve in this PR, but it's very confusing that there's `self._on_exception` and `self.on_exception` which do completely different things. 

my preference is that we remove the trainer callback hook mixin so that callback hooks aren't automatically re-exposed on the trainer this way. https://github.com/PyTorchLightning/pytorch-lightning/blob/46b00a7de44d226a3da22b258c74cc2b3cd92bec/pytorch_lightning/trainer/callback_hook.py#L27

but this is something we'll need to follow up on separately

File: tests/deprecated_api/test_remove_1-7.py [hunk 2]
```diff
@@ -116,3 +116,26 @@ def test_v1_7_0_deprecated_on_train_dataloader(tmpdir):
 def test_v1_7_0_test_tube_logger(_, tmpdir):
     with pytest.deprecated_call(match="The TestTubeLogger is deprecated since v1.5 and will be removed in v1.7"):
         _ = TestTubeLogger(tmpdir)
+
+
+def test_v1_7_0_on_interrupt(tmpdir):
+    class HandleInterruptCallback(Callback):
+        def on_keyboard_interrupt(self, trainer, pl_module):
+            self.log("keyboard interrupt")
```
Comment (ananthsub) [inline tests/deprecated_api/test_remove_1-7.py:124]:
```suggestion
            print("keyboard interrupt")
```

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -873,6 +873,52 @@ def on_keyboard_interrupt(self, trainer, pl_module):
     assert isinstance(handle_interrupt_callback.exc_info[1], KeyboardInterrupt)
 
 
+def test_on_exception_hook(tmpdir):
+    """Test the on_exception callback hook."""
+
+    model = EvalModelTemplate()
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:879]: `EvalModelTemplate` is quite legacy. I believe most of the tests have switched to the BoringModel instead.

@Borda  - potentially another project for Code Health could be to remove the dependency on EvalModelTemplate in favor of either mocked lightning modules or the boring model for integration/end-to-end tests. this could also be something new contributors could help with, similar to typing. what do you think?

**Review by ananthsub — COMMENTED** [review]
To keep the PRs single-focused, I'd recommend splitting this into:
1. 1 PR to add the `on_exception` hook
2. 1 PR to deprecate `on_keyboard_interrupt` in favor of `on_exception`

**Comment by daniellepintz** [comment]
thanks for the review @ananthsub! I separated the PRs with https://github.com/daniellepintz/pytorch-lightning/pull/1

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:518]: yeah I completely agree about _on_exception. Not sure if I'm following you - do you mean remove TrainerCallbackHookMixin and bring all the callback hooks to trainer.py, and then merge _on_exception and on_exception in the trainer?

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -873,6 +873,52 @@ def on_keyboard_interrupt(self, trainer, pl_module):
     assert isinstance(handle_interrupt_callback.exc_info[1], KeyboardInterrupt)
 
 
+def test_on_exception_hook(tmpdir):
+    """Test the on_exception callback hook."""
```
Comment (awaelchli) [inline tests/trainer/test_trainer.py:877]: can we combine this and the above test?
It's not the prettiest test I know (I wrote this)

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (daniellepintz) [reply tests/trainer/test_trainer.py:838]: Done!

**Review by ananthsub — COMMENTED** [review]
please check the test failures. we'll need to add this hook to the Lambda Callback as well: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/callbacks/lambda_function.py

**Comment by daniellepintz** [comment]
@ananthsub tests are fixed now!

**Review by awaelchli — APPROVED** [review]
great stuff!

File: pytorch_lightning/callbacks/base.py [hunk 5]
```diff
@@ -267,6 +267,10 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
         """Called when the training is interrupted by ``KeyboardInterrupt``."""
         pass
 
+    def on_exception(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", exception: BaseException) -> None:
+        """Called when any trainer execution is interrupted by an exception."""
```
Comment (Borda) [inline pytorch_lightning/callbacks/base.py:271]: do we keep it for any exception regardless of train/valid/test or during init?

File: pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -500,20 +500,22 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
...
             return trainer_fn(*args, **kwargs)
-        except KeyboardInterrupt:
+        except KeyboardInterrupt as exception:
             rank_zero_warn("Detected KeyboardInterrupt, attempting graceful shutdown...")
             # user could press Ctrl+c many times... only shutdown once
             if not self.interrupted:
                 self.state.status = TrainerStatus.INTERRUPTED
                 self.on_keyboard_interrupt()
-        except BaseException:
+                self.on_exception(exception)
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:509]: why here when it is already coved by the `KeyboardInterrupt`?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:509]: good question, the idea is to deprecate on_keyboard_interrupt in the next PR https://github.com/daniellepintz/pytorch-lightning/pull/1 so this would cover the case where someone wants to add a hook for on_exception right now, which should work for all exceptions including KeyboardInterrupt

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (daniellepintz) [reply pytorch_lightning/callbacks/base.py:271]: not sure I fully understand the question. This will be called for the main trainer function entry points (fit, validate, test, predict) - Currently it won't be called for trainer init. However, we could extend error handling to init as well, but that would be in a separate PR

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (Borda) [reply pytorch_lightning/trainer/trainer.py:509]: if so, ten deprecate it in the same PR

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/callbacks/base.py:271]: Is it an exception only for the train loop, or also the validation loop?
if so how callback can differentiate from which loop the exception came? so add it as a callback argument?

**Review by Borda — DISMISSED** [review]
need clarification on deprecation other interrupt and better spec loop from which the exception is coming from

File: pytorch_lightning/trainer/trainer.py [hunk 7]
```diff
@@ -500,20 +500,22 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
...
+        except BaseException as exception:
             self.state.status = TrainerStatus.INTERRUPTED
             if distributed_available() and self.world_size > 1:
                 # try syncing remaing processes, kill otherwise
                 self.training_type_plugin.reconciliate_processes(traceback.format_exc())
             self._on_exception()
             # reset bookkeeping
             self.state.stage = None
+            self.on_exception(exception)
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:518]: quick question: should the `on_expection raise exception` by default ? so users can prevent this behaviour if they wish.

File: tests/trainer/test_trainer.py [hunk 8]
```diff
@@ -834,10 +834,10 @@ def on_after_backward(self):
     assert not torch.isfinite(params).all()
 
 
-def test_trainer_interrupted_flag(tmpdir):
-    """Test the flag denoting that a user interrupted training."""
+def test_on_exception_hook(tmpdir):
+    """Test the on_exception callback hook and the trainer interrupted flag."""
 
-    model = EvalModelTemplate()
+    model = BoringModel()
```
Comment (tchaton) [inline tests/trainer/test_trainer.py:840]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM ! small question.

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (daniellepintz) [reply pytorch_lightning/callbacks/base.py:271]: yes it works for train loop and validation loop. you're right currently the callback can't differentiate from which loop it came. but i'm not sure if that is a requested feature. I'd suggest we merge the simple version for now, and if users request that additional feature we can add it. wdyt? do others have thoughts on this? @tchaton @ananthsub

File: pytorch_lightning/trainer/trainer.py (hunk 6 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:509]: I initially had it in this PR, but @ananthsub recommended I move it out in https://github.com/PyTorchLightning/pytorch-lightning/pull/9183#pullrequestreview-741080873 . I can merge it back if that is preferred

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:518]: that's a good idea! the only issue I see is currently for KeyboardInterrupt we don't re raise the exception so we would have to add extra logic for that. Or is it ok to raise the exception for KeyboardInterrupt as well? not sure why it's not the case currently

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:518]: For KeyboardInterrupt, I guess it won't really matter as the user know he is triggering the failure.

Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:518]: good point. sg will add it!

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:518]: Great !

Comment (daniellepintz) [reply pytorch_lightning/trainer/trainer.py:518]: I am running into an issue where if I raise the exception here: https://github.com/PyTorchLightning/pytorch-lightning/pull/9183/files#diff-0b8e28002318107af9756a2140f96a4e6165b5034a6e7ca8a551607ba39f145dR270

if there are multiple callbacks, and the first one uses default `on_exception` it will raise the error, before it can get to the other callbacks which might have a custom `on_exception`. do you know any way around this?

File: pytorch_lightning/callbacks/base.py (hunk 5 above)
Comment (ananthsub) [reply pytorch_lightning/callbacks/base.py:271]: great point @Borda !

@daniellepintz i think we should address this in the issue to ensure the design fully covers:
- When exactly is this callback called during the exception handling
- At that point, what metadata is available on the trainer object? This would inform what metadata ought to be passed through the hook
- Should end users be able to swallow the exceptions or only be able to re-raise a different exception?
- The relationship between this and https://github.com/PyTorchLightning/pytorch-lightning/issues/8895

Comment (daniellepintz) [reply pytorch_lightning/callbacks/base.py:271]: oh, I just realized the user can access the trainer state with `trainer.state.stage`. We pass the whole trainer object through the hook so the user has access to all the metadata.

regarding when this callback is called during exception handling, I think the current placement makes sense but lmk if you think it should be moved. and I think end users should be able to swallow the exceptions as discussed with @tchaton below, but I ran into an issue implementing it as I mention below

File: pytorch_lightning/trainer/trainer.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:518]: Ok, let's investigate this in another PR if we do :)
